text/0000-array-private-classes

```diff
@@ -0,0 +1,198 @@
+classes, mainly in tests. Here are the files that must be changed:
+
+* `packages/builtin/array.pony` as shown above
+* `packages/itertools/iter.pony` in function `cycle`
+* `packages/collections/heap.pony` in function `values`
+* `packages/collection/builtin/_test.pony` in class `_TestArrayValuesRewind`
+* `packages/collections/list.pony`
+* `packages/collections/map.pony`
+* `packages/collections/persistent/map.pony`
+* `packages/collections/persistent/vec.pony`
+* `packages/collections/set.pony`
+* `test/libponyc/util.cc` to change the name of the class to `_ArrayValues`
+
+# How We Teach This
+
+This change keeps the code compatible in the vast majority of cases. When client
+classes are defining objects of these now private types, the reason is usually
+to get access to the function `rewind` that was not defined in `Iterator`. By
+adding the interface `RewindableIterator`, client code can easily be adapted,
+replacing `ArrayValues[A]` by `RewindableIterator[A]`.
+
+Also, client code generally uses these functions to iterate on the returned types
+and does not try to access the iterator directly but is interested by the iterated
+items. When client code refers to the iterator type, that's generally useless and
+the code can be rewritten to be made shorter and more future proof.
+
+A [search on Github Pony code](https://github.com/search?q=%22ArrayValues%22+language%3APony&type=code)
+finds 24 files using the class `ArrayValues`, of which 6 are copies of `array.pony` file.
+
+For instance, in
+[xml2xpath.pony](https://github.com/redvers/pony-libxml2/blob/bbca5d98d48854bfec2c6ee110220873ecc4df34/pony-libxml2/xml2xpath.pony#L41),
+the code can be changed from
+
+```pony
+  fun values(): ArrayValues[Xml2node, this->Array[Xml2node]]^ ? =>
+    if (allocated) then
+      ArrayValues[Xml2node, this->Array[Xml2node]](nodearray)
+    else
+      error
+    end
+```
+
+to
+
+```pony
+  fun values(): RewindableIterator[Xml2node]^ ? =>
+    if (allocated) then
+      nodearray.values()
+    else
+      error
+    end
+```
+
+In this sample, the developer was not really concerned by the type of the iterator
+but that the `values` function must return an `RewindableIterator` over `Xml2node`.
+The new version makes the code simpler to understand.
+
+This change in `array.pony` and other collections will break such code but it
+can be easily adapted to use the new API. And it will make the standard library
+easier to learn by reducing the number of public types.
+
+# How We Test This
+
+Pony tests must continue to pass.
+
+# Drawbacks
+
+As said, some client's code must me adapted when using these classes. As these
+classes are just concreted implementations of `Iterator` by collection classes,
+their use in client code is limited and the code can very easily be changed.
+
+# Alternatives
+
+Stay as is. Continue the
+[discussion on Zulip](https://ponylang.zulipchat.com/#narrow/stream/189959-RFCs/topic/Make.20Array.20iterators.20private).
```

Comment:
If you update the RFC to also include normalizing `rewind` usage, then I think an alternative would be to keep the concrete classes but augment where it makes sense to include `rewind`. If you do include the `rewind` additions, then I think you should add the alternative here.

I believe that you also discussed previously a different way of implementing the RewindableIterface- I think it is worth adding that as an alternative here.

Reply:
> I think an alternative would be to keep the concrete classes

You didn't show an example where having these classes is justified outside the collections code.

This alternative reminds me of Sun, ten years ago before being bought by Oracle, keeping some buggy classes in the Java stdlib because some of its customers were using them and having built software upon non-documented (understand buggy) behaviour. But Sun had tens of millions of Java users, many of them paying for a Java license or support, and Java has a [deprecation procedure](https://docs.oracle.com/javase/7/docs/technotes/guides/javadoc/deprecation/deprecation.html) to allow API transition between language versions (since version 1.1). None of these conditions applies to Pony yet... Is it worth adding an alternative not supported by concrete proofs?

Reply:
I'm not asking for a change that will break some code. You are. I don't need to justify anything. I presenting you with items that will make your RFC that I am not at the moment in favor of stronger. You do not have to make your RFC stronger, that is up to you. Despite my not being in favor of this RFC, I've been spending my free time helping you make it more convincing, I don't have to do that.